src/libmongoc/src/mongoc/mongoc-shared.h

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2021 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MONGOC_SHARED_H
+#define MONGOC_SHARED_H
+
+#include <stddef.h>
+
+struct _mongoc_shared_ptr_aux;
+
+/**
+ * @brief A ref-counted thread-safe shared pointer to arbitrary data
+ */
+typedef struct mongoc_shared_ptr {
+   /** Pointed-to data */
+   void *ptr;
+   /** Auxilary book-keeping. Do not touch. */
+   struct _mongoc_shared_ptr_aux *_aux;
+} mongoc_shared_ptr;
```

Comment:
To check, can a pointer to the same `mongoc_shared_ptr` be safely shared between threads? Or do users of `mongoc_shared_ptr` only be sharing copies of the struct between threads?

Based on the comment above `mongoc_shared_ptr_rebind_atomic` I believe a pointer to `mongoc_shared_ptr` can be shared between threads, as long as the `*_atomic` functions are used as appropriate.

Reply:
There's three different objects when dealing with a shared_ptr: A `shared_ptr` instance, the shared state (stored in `_aux`), and the pointed-to data. Only the shared state is thread-safe. A `shared_ptr` object is not itself thread-safe, which is why we must duplicate the shared_ptr via `take_atomic(&p)` if `p` is accessible between threads and one of those threads might rebind `p` (as is the case with `topology->_shared_descr_`).

In the case of `_shared_descr_`, the pointee is `const`, and thus requires no synchronization as no other thread can modify it while we use it (as long as no one tries anything tricky). The only shared mutable state is the `_shared_descr_` object itself, thus requiring the synchronization.

Reply:
Thank you for the explanation. I now see that the `mongoc_shared_ptr` struct cannot safely be accessed between threads, and must be copied.